### ISSUE TITLE
http-request replace-path does not work for websocket requests

### ISSUE THREAD

**Issue by nebilbenmabrouk** [haproxy/haproxy#829]
## Detailed description of the problem
I have multiple websocket backends (namely, multiple JupyterLab instances), 
- JupyterLab  instance 1 running in:  http://jupyter-1-host:30001/
- JupyterLab  instance 2 running in:  http://jupyter-2-host:30002/

I add some query parameters in the websocket requests (of each instance) to know which backend to use for each request. For instance, I tell JupyterLab instances to use as websocket URLs: 
- `--NotebookApp.websocket_url=ws://haproxy-host:8888/?endpoint=jupyter-instance-1` (for JupyterLab  instance 1)
- `--NotebookApp.websocket_url=ws://haproxy-host:8888/?endpoint=jupyter-instance-2` (for JupyterLab  instance 2)

In the file haproxy.cfg, in the backend configurations, I added a **http-request replace-path** instruction in order to remove the query param   _endpoint=jupyter-instance-1_ before sending the websocket request to the backend.
The instruction is:
`http-request replace-path /(.*)?endpoint=jupyter-instance-1(.*) /\1\2`  

This instruction works fine for HTTP requests, but NOT for WEBSOCKET requests (starting with ws or wss)

## Expected behavior
HAProxy should send websocket requests in the form of: `ws://jupyter-1-host:30001/`

## Actual behavior
HAProxy sends websocket requests in the form of: `ws://jupyter-1-host:30001/?endpoint=jupyter-instance-1`
which prevents JupyterLab websockets from working.

## Steps to reproduce the behavior
1. Configure two instances of JupyterLab to use the following parameters, respectively. 
`--NotebookApp.websocket_url=ws://haproxy-host:8888/?endpoint=jupyter-instance-1`
`--NotebookApp.websocket_url=ws://haproxy-host:8888/?endpoint=jupyter-instance-2`
2. Run the two instances of JupyterLab
3. Add the following Frontend in the file haproxy.cfg
```
frontend JupyterServices
    bind *:8888
    use_backend jupyter-1   if  { hdr_beg(Host) -i ws } { url_sub -i endpoint=jupyter-instance-1 }
    use_backend jupyter-2   if  { hdr_beg(Host) -i ws } { url_sub -i endpoint=jupyter-instance-2 }
```
4. Add the following Backends in the file haproxy.cfg
```
backend jupyter-1
    http-request replace-path /(.*)?endpoint=jupyter-instance-1(.*) /\1\2
    http-request set-header Upgrade websocket
    http-request set-header Connection Upgrade
    server jupyter-instance-1  jupyter-1-host:30001

backend jupyter-2
    http-request replace-path /(.*)?endpoint=jupyter-instance-2(.*) /\1\2
    http-request set-header Upgrade websocket
    http-request set-header Connection Upgrade
    server jupyter-instance-2  jupyter-2-host:30002
```
5. Start HAProxy
6. In the browser go to http://haproxy-host:8888/?endpoint=jupyter-instance-1
Then open the webtools of the browser to see the details of the broken websocket requests.

## Do you have any idea what may have caused this?
http-request replace-path works only for URIs starting with **http** or **https**, but not for URIs starting with **ws** or **wss**

## Do you have an idea how to solve the issue?
Consider websocket requests for **http-request** instructions.

## What is your configuration?
This the main part of the configuration
```
frontend JupyterServices
    mode http
    bind *:8888
    use_backend jupyter-1   if  { hdr_beg(Host) -i ws } { url_sub -i endpoint=jupyter-instance-1 }
    use_backend jupyter-2   if  { hdr_beg(Host) -i ws } { url_sub -i endpoint=jupyter-instance-2 }

backend jupyter-1
    http-request replace-path /(.*)?endpoint=jupyter-instance-1(.*) /\1\2
    http-request set-header Upgrade websocket
    http-request set-header Connection Upgrade
    server jupyter-instance-1  jupyter-1-host:30001

backend jupyter-2
    http-request replace-path /(.*)?endpoint=jupyter-instance-2(.*) /\1\2
    http-request set-header Upgrade websocket
    http-request set-header Connection Upgrade
    server jupyter-instance-2  jupyter-2-host:30002
```

**Comment by git001** [comment]
What's the haproxy version?

**Comment by nebilbenmabrouk** [comment]
> What's the haproxy version?

Thank you for your quick reply. I use HAProxy version 2.3

**Comment by wtarreau** [comment]
I am confused, your proposed config has nothing to do with your description. You're filtering on host names starting with "ws" while that's not what you're doing in your description.

Also, asking haproxy to insert the Upgrade header itself on behalf of the client is wrong: you're asking haproxy and the server to upgrade their connection to websocket but the connection between haproxy and the client to remain HTTP, that cannot work, the client will get a 101 that is unexpected.

Please provide your request logs and/or run haproxy in debug mode (-d) to actually see your requests and know what to match against and what you want to adjust.

**Comment by TimWolla** [comment]
> I use HAProxy version 2.3

You should not do this. 2.3 is a development version. Use HAProxy 2.2 (2.2.2 is current as of now) from here: https://www.haproxy.org/

**Comment by nebilbenmabrouk** [comment]
> > I use HAProxy version 2.3
> 
> You should not do this. 2.3 is a development version. Use HAProxy 2.2 (2.2.2 is current as of now) from here: https://www.haproxy.org/

I tried also with the version 2.2.2 and it does not work. 
The problem is that **http-request replace-path** completely ignores websocket requests

**Comment by nebilbenmabrouk** [comment]
> I am confused, your proposed config has nothing to do with your description. You're filtering on host names starting with "ws" while that's not what you're doing in your description.
=> Filtering on host headers starting with "ws" is a recommended way to detect websocket requests, the other recommended way is to use **{ hdr(Upgrade) -i websocket }**, I tried both and they work fine, the problem is related to the **http-request replace-path** instruction.

> Also, asking haproxy to insert the Upgrade header itself on behalf of the client is wrong: you're asking haproxy and the server to upgrade their connection to websocket but the connection between haproxy and the client to remain HTTP, that cannot work, the client will get a 101 that is unexpected.
=> Inserting websocket headers (including the Upgrade header) is mandatory to make Jupyter work. Otherwise, HAProxy will send a HTTP request to the backend instead of a WebSocket request. When I have a single Jupyter instance and I do not need the **http-request replace-path** instruction, all my configuration works fine (please refer to the recommendations for Jupyter to work behind a reverse proxy).

> Please provide your request logs and/or run haproxy in debug mode (-d) to actually see your requests and know what to match against and what you want to adjust.
=> Here is the websocket request that does not work (note ?endpoint=jupyter-instance-1): 
ws://haproxy-host:8888/?endpoint=jupyter-instance-1/api/kernels/7539fe92-6c76-4d9d-8582-ea93e5207a84/channels?session_id=66260bab-94b2-4f36-9774-031cf1ee3144&token=toto
The correct one that should be sent is: 
ws://haproxy-host:8888/api/kernels/7539fe92-6c76-4d9d-8582-ea93e5207a84/channels?session_id=66260bab-94b2-4f36-9774-031cf1ee3144&token=toto

**Comment by git001** [comment]
Well the WebSocket communication is bypassed by the http engine afaik, that's the reason why HAProxy can't change the Requests **inside** WebSocket *after* connection between client and server is established.

That's also the reason why [timeout tunnel](http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#timeout%20tunnel) is relevant for WebSocket because HAProxy switches to tunnel mode.

Quote from doc
> In HTTP, this timeout is used when a connection is upgraded (e.g.
when switching to the WebSocket protocol, or forwarding a CONNECT request
to a proxy), or after the first response when no keepalive/close option is
specified.

Maybe you take a deeper look into the following descriptions [WebSocket](https://en.wikipedia.org/wiki/WebSocket) and [How Do Websockets Work?](https://sookocheff.com/post/networking/how-do-websockets-work/).

Especially this block.

> Once the connection is established, the client and server can send WebSocket data or text frames back and forth in full-duplex mode. The data is minimally framed, with a small header followed by payload.[36] WebSocket transmissions are described as "messages", where a single message can optionally be split across several data frames. This can allow for sending of messages where initial data is available but the complete length of the message is unknown (it sends one data frame after another until the end is reached and marked with the FIN bit). With extensions to the protocol, this can also be used for multiplexing several streams simultaneously (for instance to avoid monopolizing use of a socket for a single large payload).[37]

**Comment by nebilbenmabrouk** [comment]
> Well the WebSocket communication is bypassed by the http engine afaik, that's the reason why HAProxy can't change the Requests **inside** WebSocket _after_ connection between client and server is established.
=> I DO NOT want to "change the Requests **inside** WebSocket _after_ connection between client and server is established".
**The problem occurs before the connection is established because the URI of the websocket request is wrong** 

> That's also the reason why [timeout tunnel](http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#timeout%20tunnel) is relevant for WebSocket because HAProxy switches to tunnel mode.
=> I configured the parameter **timeout tunnel**, and when I have a single backend, everything works fine.

> Quote from doc
> 
> > In HTTP, this timeout is used when a connection is upgraded (e.g.
> > when switching to the WebSocket protocol, or forwarding a CONNECT request
> > to a proxy), or after the first response when no keepalive/close option is
> > specified.
> 
> Maybe you take a deeper look into the following descriptions [WebSocket](https://en.wikipedia.org/wiki/WebSocket) and [How Do Websockets Work?](https://sookocheff.com/post/networking/how-do-websockets-work/).

> Especially this block.
> 
> > Once the connection is established, the client and server can send WebSocket data or text frames back and forth in full-duplex mode. The data is minimally framed, with a small header followed by payload.[36] WebSocket transmissions are described as "messages", where a single message can optionally be split across several data frames. This can allow for sending of messages where initial data is available but the complete length of the message is unknown (it sends one data frame after another until the end is reached and marked with the FIN bit). With extensions to the protocol, this can also be used for multiplexing several streams simultaneously (for instance to avoid monopolizing use of a socket for a single large payload).[37]

=> I know how websockets work, and again the **problem occurs before the connection is established because the URI of the websocket request is wrong** 

My QUESTION is : 
1- Does the instruction **http-request replace-path** applies for websocket requests (since websocket requests are HTTP requests with specific scheme and headers), so I consider the issue as a bug, or
2- It does not apply for websocket requests, so I change this issue to a feature request.

Thank you in advance for your help

**Comment by lukastribus** [comment]
To understand whether or not replace-path did it's job (while still in HTTP mode), you must not look in the browsers developer tools, those will always show the original path, because you have configured haproxy to rewrite the request towards the server, not redirect the client somewhere else.

You need to look in the server logs or capture the haproxy -> backend server traffic instead.

**Comment by wtarreau** [comment]
[ please don't break quotes in replies as this becomes hard to follow ]

> > I am confused, your proposed config has nothing to do with your description. You're filtering on host names starting with "ws" while that's not what you're doing in your description.

> Filtering on host headers starting with "ws" is a recommended way to detect websocket requests, the other recommended way is to use **{ hdr(Upgrade) -i websocket }**, I tried both and they work fine, the problem is related to the **http-request replace-path** instruction.

OK so you're not directing your browser to "haproxy-host:8888" but "ws-something" instead, right ? Sorry if this sounds obvious to you, but as the method you're suggesting to reproduce the issue does not correspond to your configuration, it's particularly difficult to understand where the problem could be.

> > Also, asking haproxy to insert the Upgrade header itself on behalf of the client is wrong: you're asking haproxy and the server to upgrade their connection to websocket but the connection between haproxy and the client to remain HTTP, that cannot work, the client will get a 101 that is unexpected.

> Inserting websocket headers (including the Upgrade header) is mandatory to make Jupyter work. Otherwise, HAProxy will send a HTTP request to the backend instead of a WebSocket request.

No, I disagree. This header is added by your browser and should be relayed by haproxy. If you have to add it, there's definitely something else which is wrong.

> When I have a single Jupyter instance and I do not need the **http-request replace-path** instruction, all my configuration works fine (please refer to the recommendations for Jupyter to work behind a reverse proxy).

The problem is that your configuration shows that many things don't work as they should, so the replace-path point cannot be troubleshooted until you have a properly working configuration. It could be a side effect of all the anomalies around or could be a colateral victim of a same issue.

> > Please provide your request logs and/or run haproxy in debug mode (-d) to actually see your requests and know what to match against and what you want to adjust.

> Here is the websocket request that does not work (note ?endpoint=jupyter-instance-1):
> ws://haproxy-host:8888/?endpoint=jupyter-instance-1/api/kernels/7539fe92-6c76-4d9d-8582-ea93e5207a84/channels?session_id=66260bab-94b2-4f36-9774-031cf1ee3144&token=toto

So you see, here you're having `haproxy-host` again, not `ws-something`. Based on your configuration, this will not go to any of the two backends above and thus the replace rule will not be executed. So please at least provide valid information to save people from searching in the wrong direction. Have you at least checked in your logs that your request pass through the correct backend ?

If you have enabled a `stats socket`, you can get full runtime traces both on input and output this way:
```
(echo "trace h1 sink buf0; trace h1 level developer; trace h1 verbosity complete; trace h1 start now; show events buf0 -w -n" ; cat ) | socat - /path/to/stats/socket
```

**Comment by capflam** [comment]
I found a bug in the `replace-path` action. The documentation specifies the path contains the query string if any
is present. But for now, the replacement is only applied on the path part, excluding the query-string.

Note that your regular expression will not do what you expect. You must escape the question mark :

```
http-request replace-path /(.*)\?endpoint=jupyter-instance-1(.*) /\1\2
```

Then, independently on this bug, the `replace-path` action only applies on the path part. The scheme part is never evaluated. But, from HAProxy point of view (or an HTTP server one), the `ws` (or `wss`) scheme does not mean anything. The client always sends valid HTTP request based on infos extracted from the WebSocket URI. And the `Host` header does not contain the scheme. Thus, when your browser opens a WebSocket connection using your URI, it sends this kind of request:

```http
GET /?endpoint=jupyter-instance-1/api/kernels/7539fe92-6c76-4d9d-8582-ea93e5207a84/channels?session_id=66260bab-94b2-4f36-9774-031cf1ee3144&token=toto HTTP/1.1
host: haproxy-host:8888
upgrade: websocket
connection: Upgrade
sec-websocket-key: ...
...
```

Note that as Willy said, you must not configure HAProxy to set `Upgrade` and `Connection` headers. It is the client responsibility to initiate a WebSocket handshake.  And based on your configuration snippet, your `use_backend` rules are definitely wrong, except if `haproxy-host` is in reality starting with `ws`.

I will fix the `replace-path` bug. In the mean time, as a workaround, you may use `replace-uri` action. Most of time, in HTTP/1.1, HTTP clients send requests using the origin-form of the request target (i.e, `absolute-path [ "?" query ]`)

**Comment by capflam** [comment]
It should be fixed in upstream now.

**Comment by nebilbenmabrouk** [comment]
> I found a bug in the `replace-path` action. The documentation specifies the path contains the query string if any
> is present. But for now, the replacement is only applied on the path part, excluding the query-string.
> 
> Note that your regular expression will not do what you expect. You must escape the question mark :
> 
> ```
> http-request replace-path /(.*)\?endpoint=jupyter-instance-1(.*) /\1\2
> ```
> 
> Then, independently on this bug, the `replace-path` action only applies on the path part. The scheme part is never evaluated. But, from HAProxy point of view (or an HTTP server one), the `ws` (or `wss`) scheme does not mean anything. The client always sends valid HTTP request based on infos extracted from the WebSocket URI. And the `Host` header does not contain the scheme. Thus, when your browser opens a WebSocket connection using your URI, it sends this kind of request:
> 
> ```httpspec
> GET /?endpoint=jupyter-instance-1/api/kernels/7539fe92-6c76-4d9d-8582-ea93e5207a84/channels?session_id=66260bab-94b2-4f36-9774-031cf1ee3144&token=toto HTTP/1.1
> host: haproxy-host:8888
> upgrade: websocket
> connection: Upgrade
> sec-websocket-key: ...
> ...
> ```
> 
> Note that as Willy said, you must not configure HAProxy to set `Upgrade` and `Connection` headers. It is the client responsibility to initiate a WebSocket handshake. And based on your configuration snippet, your `use_backend` rules are definitely wrong, except if `haproxy-host` is in reality starting with `ws`.
> 
> I will fix the `replace-path` bug. In the mean time, as a workaround, you may use `replace-uri` action. Most of time, in HTTP/1.1, HTTP clients send requests using the origin-form of the request target (i.e, `absolute-path [ "?" query ]`)



> It should be fixed in upstream now.


Thanks a lot for your agile answers. I'll apply your recommendations and see if it will solve the issue. Otherwise I'll upgrade HAProxy.

**Comment by wtarreau** [comment]
Actually I think we have some serious inconsistencies in various actions and sample fetch functions which have caused this confusion, and I'm not fond of backporting this to 2.0 as mentioned in the commit. Indeed:

- the `path` sample fetch function works exactly as the RFC, which means that it stops before the question mark delimiting the query string
- the `set-path` action also acts according to above, which means that it will only rewrite the path part of the request and will preserve the query string.
- the `replace-path` action *used* to do that (except that it was documented differently) and now it acts differently.

I find it extremely confusing to think that `set-path` doesn't clear the query string, that `set-path %[path]` is an identity, while replace-path on the same values will drop the query string. Further, by changing it there is a significant risk of breaking existing setups. This action was added relatively recently, in 2.2, so we can expect that the few existing deployments work as expected (if any).

I'd rather instead add a new action (and possibly even sample fetches etc) that we could call "something" (request/full-path/pathq/pq ?) that would contain both the path and the query string and keep replace-path on the well-defined "path" component (i.e. before query string). Because quite frankly, I expect way more problems *after* the current fix than before.

**Comment by wtarreau** [comment]
Thinking about it, "pathq" isn't that bad because it would appear immediately after "path" in the doc, making its availability obvious to any user. So we could have a "pathq" sample fetch returning path+optional query string, and a replace-pathq action to replace them. And even possibly a set-pathq one if we find it useful. What do you think ?

**Comment by capflam** [comment]
It is a good idea. I agree there is an inconsistency. Purpose of my patch was to respect the documentation. There are even examples in `replace-path` documentation manipulating the query-string. Thus, if we don't backport it to 2.0, we must change the documentation. Or wait to add `replace-pathq` action and backport everything as far as 2.0.

**Comment by TimWolla** [comment]
> Further, by changing it there is a significant risk of breaking existing setups. This action was added relatively recently, in 2.2, so we can expect that the few existing deployments work as expected (if any).

Agreed. I would say that for behavior that can _sensible_ be argued either way we should assume that the bug is in the documentation, not in the code. In fact I started checking all my configurations to see whether I am using `replace-path` anywhere to see whether it would break for me. Not everyone stays on top of all the changes happening and by using distro packages (or e.g. Vincent's private repository) they'll receive any patch updates automatically.

**Comment by wtarreau** [comment]
@capflam yes I agree that the examples are further confusing, because if you look closely at them, they will all work in either case just because the query string is either optional or not changed! This is particularly nasty! I'd rather revert the change for now to keep things consistent between versions and use that code to create replace-pathq once we have time for this in fact. In the short term we should definitely address that stupidity in the doc. And I'm the one who did it, yes, I know :-/

**Comment by capflam** [comment]
FYI, I pushed the fix. `replace-path` is now replacing the path only and documented this way. In addition, there are 2 new actions, `set-pathq` and `replace-pathq`, to manipulate the path including the query string. And `pathq` sample fetch has been added too.

I will do all required backports soon. @nebilbenmabrouk, once backported in 2.2, you will have to use `replace-pathq` to do what you want.
